File: src/main/java/no/nav/pensjonbrevdata/helpers/DokumentmalGenerators.java
package no.nav.pensjonbrevdata.helpers;

import java.io.File;
import java.io.IOException;
import java.util.function.Function;

public class DokumentmalGenerators {

    public static Function<String, String> dokumentmalGenerator = dokumentmalBuilder("dokumentmal");
    public static Function<String, String> fellesmalGenerator = dokumentmalBuilder("felles");

    private DokumentmalGenerators() { }

    private static Function<String, String> dokumentmalBuilder(String dokumentMalType){
        return dokumentmalId -> {
            try {
                return new XsdFileReader().read("xsd" + File.separator + dokumentMalType + File.separator + dokumentmalId + ".xsd");
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        };
    }
}

File: src/main/java/no/nav/pensjonbrevdata/helpers/DokumentmalGenerators.kt
package no.nav.pensjonbrevdata.helpers

import java.io.File

object DokumentmalGenerators {
    @JvmField
    val dokumentmalGenerator: (String) -> String = dokumentmalBuilder("dokumentmal")

    @JvmField
    val fellesmalGenerator: (String) -> String = dokumentmalBuilder("felles")

    private fun dokumentmalBuilder(dokumentMalType: String): (String) -> String = { dokumentmalId: String ->
        read("xsd" + File.separator + dokumentMalType + File.separator + dokumentmalId + ".xsd")
    }

    private fun read(resourcePath: String): String = javaClass.getResource("/$resourcePath").readText(Charsets.UTF_8)
}
